syncopy/tests/test_mne_conv.py

```diff
@@ -0,0 +1,158 @@
+
+
+import syncopy as spy
+from syncopy.synthdata.analog import white_noise
+import numpy as np
+import pytest
+
+has_mne = False
+try:
+    import mne
+    has_mne = True
+except ImportError:
+    pass
+
+skip_no_mne = pytest.mark.skipif(not has_mne, reason="MNE Python not installed")
+
+class TestSpyToMNE():
+
+    numChannels = 64
+    numTrials = 5
+    numSamples = 1000
+    adata = white_noise(nTrials = numTrials, nChannels=numChannels, nSamples=numSamples)
+
+    @skip_no_mne
+    def test_spy_analog_raw_to_mne(self):
+        """
+        Test conversion of spy.AnalogData to MNE RawArray.
+
+        This uses raw data, i.e., data without trial definition.
+        """
+
+        adata = self.adata
+        # Convert to MNE RawArray
+        ar = spy.io.mne_conv.raw_adata_to_mne(adata)
+
+        assert type(ar) == mne.io.RawArray
+        # Check that the data is the same
+        assert np.allclose((adata.data[()]).T, ar.get_data())
+        # Check that the channel names are the same
+        assert all(adata.channel == ar.ch_names)
+        # Check that the sampling rate is the same
+        assert adata.samplerate == ar.info['sfreq']
+
+    @skip_no_mne
+    def test_mne_rawarray_to_spy_analog(self):
+        """
+        Test conversion of mne.io.RawArray to spy.AnalogData.
+
+        This uses raw data, i.e., data without trial definition.
+        """
+
+        adata = self.adata
+        # Convert to MNE RawArray
+        ar = spy.io.mne_conv.raw_adata_to_mne(adata)
+        # Now convert back to AnalogData
+        adata2 = spy.io.mne_conv.raw_mne_to_adata(ar)
+        assert type(adata2) == spy.AnalogData
+        assert all(adata.channel == adata2.channel)
+        assert np.allclose(adata.data, adata2.data)
+        assert adata.samplerate == adata2.samplerate
+
+    @skip_no_mne
+    def test_tldata_to_mne_with_TimeLockData(self):
+        """
+        Test conversion of spy.TimeLockData to mne.EpochsArray.
+
+        This uses epoched data, i.e., data with trial definition and trials of identical length (and offset), i.e., timelocked data.
+        """
+        adata = self.adata
+        assert type(adata) == spy.AnalogData
+        tldata = spy.timelockanalysis(adata, latency="maxperiod")
+        assert type(tldata) == spy.TimeLockData
+        # Convert to MNE EpochsArray
+        epoched = spy.io.mne_conv.tldata_to_mne(tldata)
+        assert type(epoched) == mne.EpochsArray
+
+        # Check dimensions
+        n_times = epoched.get_data().shape[2]
+        assert n_times == self.numSamples
+        n_epochs = epoched.get_data().shape[0]
+        assert n_epochs == self.numTrials
+        n_channels = epoched.get_data().shape[1]
+
+        assert n_times == tldata.trials[0].shape[0]
+        assert n_epochs == len(tldata.trials)
```

Comment:
`epochs = trials`.. jeez all this nomenclature still is quite confusing

Reply:
myunderstanding: for MNE, there are trials (the short recording snippets, one for each repetition of a task) and epochs (subsets of recordings, which  are timelocked with respect to some event during the trial, i.e., the event happens at the same time relative to the start of the epoch, for all epochs).

In MNE, people seem to use only epochs during the analysis in many cases, as trials are something from the raw data domain that is typically not of relevance for the analysis. syncopy does not have epochs, so we have to store epochs as trials.

Reply:
Uh, ok that's even more confusing. A proper definition of what FT/Syncopy considers a trial would help for sure. Just not overly confident that we are the perfect ppl to write it 😅

Reply:
MNE has a definition of what an Epoch is (or, what their EpochsArray datastructure models) at https://mne.tools/stable/auto_tutorials/epochs/10_epochs_overview.html

Quoting the relevant part:

[Epochs](https://mne.tools/stable/generated/mne.Epochs.html#mne.Epochs) objects are a data structure for representing and analyzing equal-duration chunks of the EEG/MEG signal. [Epochs](https://mne.tools/stable/generated/mne.Epochs.html#mne.Epochs) are most often used to represent data that is time-locked to repeated experimental events (such as stimulus onsets or subject button presses), but can also be used for storing sequential or overlapping frames of a continuous signal (e.g., for analysis of resting-state activity; see [Making equally-spaced Events arrays](https://mne.tools/stable/auto_tutorials/raw/20_event_arrays.html#fixed-length-events)). Inside an [Epochs](https://mne.tools/stable/generated/mne.Epochs.html#mne.Epochs) object, the data are stored in an [array](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray) of shape (n_epochs, n_channels, n_times).

Reply:
I will try to write something, and let Diljit and Mohsen double check it.